src/goto-instrument/contracts/contracts.cpp

```diff
@@ -133,34 +137,66 @@ void code_contractst::check_apply_loop_contracts(
     parameter2history,
     loop_head->source_location(),
     mode,
-    havoc_code,
+    generated_code,
     ID_loop_entry);
 
   // Create 'loop_entry' history variables
-  insert_before_swap_and_advance(goto_function.body, loop_head, havoc_code);
+  insert_before_swap_and_advance(goto_function.body, loop_head, generated_code);
 
   // Generate: assert(invariant) just before the loop
   // We use a block scope to create a temporary assertion,
   // and immediately convert it to goto instructions.
   {
     code_assertt assertion{invariant_expr()};
     assertion.add_source_location() = loop_head->source_location();
-    converter.goto_convert(assertion, havoc_code, mode);
-    havoc_code.instructions.back().source_location_nonconst().set_comment(
+    converter.goto_convert(assertion, generated_code, mode);
+    generated_code.instructions.back().source_location_nonconst().set_comment(
       "Check loop invariant before entry");
   }
 
   // havoc the variables that may be modified
-  havoc_if_validt havoc_gen(modifies, ns);
-  havoc_gen.append_full_havoc_code(loop_head->source_location(), havoc_code);
+  modifiest modifies;
+  if(assigns.is_nil())
```

Comment:
This method is way too big already, could the assigns-handling please be factored out into a method of its own?

Reply:
+1 what he said.

Reply:
Much of the assigns clause checking is in a separate function (`check_frame_conditions`). In this function we call:

```c
  if(assigns.is_not_nil())
  {
    check_frame_conditions(
      function_name, goto_function.body, loop_head, loop_end, loop_assigns);
  }
```

and we have a few more lines to create snapshots of what's assignable:
```c
    for(const auto &car : loop_assigns.get_write_set())
    {
      auto snapshot_instructions = car.generate_snapshot_instructions();
      insert_before_swap_and_advance(
        goto_function.body, loop_head, snapshot_instructions);
    };
```

So these parts are not too big by themselves (and appear at different points within this function), but I agree that this function is growing to be too large. I would probably do this in a separate refactoring PR and pull out loop contracts to a separate `loop_contracts.cpp` file.

Reply:
I would be fine with a separate refactoring commit.